beacon-chain/forkchoice/protoarray/nodes.go

```diff
@@ -64,6 +68,15 @@ func (s *Store) updateCanonicalNodes(ctx context.Context, root [32]byte) error {
 	ctx, span := trace.StartSpan(ctx, "protoArrayForkChoice.updateCanonicalNodes")
 	defer span.End()
 
+	// Grab multiple locks, prefer grabbing write lock first since it is the most likely lock to
+	// block.
+	s.canonicalNodesLock.Lock()
+	defer s.canonicalNodesLock.Unlock()
+	s.nodesLock.RLock()
+	defer s.nodesLock.RUnlock()
+	s.nodeIndicesLock.RLock()
+	defer s.nodeIndicesLock.RUnlock()
```

Comment:
If I'm not mistaken, this code can lead to deadlock scenarios. Let's say some other function tries to acquire the same locks, but in different order:
```
s.canonicalNodesLock.Lock()
defer s.canonicalNodesLock.Unlock()
s.nodeIndicesLock.RLock()
defer s.nodeIndicesLock.RUnlock()
s.nodesLock.RLock()
defer s.nodesLock.RUnlock()
```
It might happen that `updateCanonicalNodes` acquires `nodesLock` and the other function acquires `nodeIndicesLock`. They will deadlock waiting for each other.

Reply:
Maybe we could use https://github.com/atedja/go-multilock?

Reply:
Yes, this is high risk for a deadlock. I'll take a look at alternatives, multilock seems like it would be appropriate.